packages/acm-dns-validated-cert/src/index.ts

```diff
@@ -0,0 +1,116 @@
+import * as pulumi from '@pulumi/pulumi'
+import * as aws from '@pulumi/aws'
+
+export interface ZoneNameArgs {
+    zoneName?: string
+}
+
+export interface ZoneIdArgs {
+    zoneId?: string
+}
+
+export interface CommonArgs {
+    name?: string | pulumi.Input<string>
+    subject: string | pulumi.Input<string>
+    subjectAlternativeNames?: string[]
+}
+
+type ACMCertArgs = ZoneIdArgs & ZoneNameArgs & CommonArgs
+
+export class ACMCert extends pulumi.ComponentResource {
+    public readonly certificate: aws.acm.Certificate
+    public readonly zoneId: string | pulumi.Input<string>
+    public readonly certificateValidation: aws.acm.CertificateValidation
+    public readonly certificateArn: pulumi.Output<string>
+
+    constructor(
+        name: string,
+        private args: ACMCertArgs,
+        opts?: pulumi.ResourceOptions,
+    ) {
+        super('ACMCert', name, {}, opts)
+
+        const selectedName = args.name ? args.name : name
+
+        if (args.zoneId) {
+            this.zoneId = args.zoneId
+        } else if (args.zoneName) {
+            this.zoneId = pulumi
+                .output(
+                    aws.route53.getZone(
+                        {
+                            name: args.zoneName,
+                        },
+                        {
+                            ...opts,
+                            parent: this,
+                            async: true,
+                        },
+                    ),
+                )
+                .apply((zone: aws.route53.GetZoneResult) => zone.id)
+        } else {
+            throw new Error(
+                'You must set either zoneName or zoneId when creating a new ACMCert',
+            )
+        }
+
+        this.certificate = new aws.acm.Certificate(
+            `${selectedName}-cert`,
+            {
+                domainName: args.subject,
+                subjectAlternativeNames: args.subjectAlternativeNames,
+                validationMethod: 'DNS',
+            },
+            {
+                ...opts,
+                parent: this,
+            },
+        )
+
+        const numberOfUniqueDomains = [
+            ...new Set([args.subject, ...(args.subjectAlternativeNames || [])]),
+        ].length
+
+        const certValidationRecords = Array(numberOfUniqueDomains)
+            .fill(null)
+            .map(
+                (_, i) =>
+                    new aws.route53.Record(
+                        `${selectedName}-validation-record-${i}`,
+                        {
+                            name: this.certificate.domainValidationOptions[i]
```

Comment:
The index will be wrong here if items are de-duped. 

Instead just create a `uniqueDomains` set, then map over it?

Reply:
These records need to be created dynamically, based on the validation records from the certificate, but pulumi doesn't allow that.

Flow:
1. AWS creates the certificate, but sets its status to PENDING. 
2. AWS returns a set of validation records that you must create to prove ownership of the domain.
3. AWS waits for the records to be created, and then issues the certificate (its status becomes ISSUED). At this point, the certificate is available for use in other resources.

In pulumi (and terraform), the validation records returned at step 2 are outputs, but pulumi doesn't allow creating resources in `.apply()` &mdash; in other words, the resource graph is locked in at the plan phase. You can hook up outputs and inputs to fill in the blanks in the dependency graph, but after the plan is complete, you can't add new object to the dependency graph.

Unfortunately, we need to use the output of the ACM cert to determine how many DNS records we need, but when the output is available, the dependency graph is already locked in.

Terraform has the same limitation.

As a workaround, this module infers the expected number of validation records, using the number of unique domains added to the certificate. From this, the DNS records are defined, and the validationOptions are used as inputs.

It works, though it's a bit messy.

Reply:
In other words, yes we can create uniqueDomains, but that's not the real problem. :)

Reply:
> The index will be wrong here if items are de-duped.
> 
> Instead just create a `uniqueDomains` set, then map over it?

I've refactored the code to hopefully make it a bit more obvious what's happening.